lib/puppet/network/http/api/indirected_routes.rb

```diff
@@ -109,7 +109,7 @@ def uri2indirection(http_method, uri, params)
       raise Puppet::Network::HTTP::Error::HTTPNotAuthorizedError.new(e.message)
     end
 
-    if configured_environment.nil?
+    if configured_environment.nil?  && indirection_name != "catalog"
```

Comment:
Added this so the server would not fail and return a catalog of the non-existing environment.
Could be a better approach on this

Reply:
I think we could break this up into two parts. First, ask the indirection's default terminus if it's prepared to handle an environment that doesn't exist. So add a boolean method to [Puppet::Indirection::Terminus](https://github.com/puppetlabs/puppet/blob/846e5e41d50d363e738658588c497313a59900a5/lib/puppet/indirector/terminus.rb#L146-L148), something like `require_environment?`

By default, have it return `true` and modify the `indirected_routes` to call it:

```ruby
if configured_environment.nil? && indirection.terminus.require_environment?
  raise ...
```

Then update the [`compiler` terminus](https://github.com/puppetlabs/puppet/blob/main/lib/puppet/indirector/catalog/compiler.rb) to override `require_environment?`

```ruby
def require_environment?
  false
end
```

And update its `find` method to check if the environment exists:

```ruby
def find(..)
  return nil unless Puppet.lookup(:environments).get(request.environment)
  ..
```

Returning `nil` should have the same effect as the indirected routes handler raising `HTTPNotFoundError` due to 
https://github.com/puppetlabs/puppet/blob/846e5e41d50d363e738658588c497313a59900a5/lib/puppet/network/http/api/indirected_routes.rb#L120-L123

At this point, everything should work the same as before, except for one thing. The indirected routes class deletes the [`environment` query parameter](https://github.com/puppetlabs/puppet/blob/846e5e41d50d363e738658588c497313a59900a5/lib/puppet/network/http/api/indirected_routes.rb#L61) and it is only readded if the environment exists:

https://github.com/puppetlabs/puppet/blob/846e5e41d50d363e738658588c497313a59900a5/lib/puppet/network/http/api/indirected_routes.rb#L95-L99

So maybe we only delete the parameter if `require_environment?` returns true?

Next, have the agent send a query parameter asking the server to redirect if the environment it's requesting doesn't exist. Maybe it's the same parameter as what I was suggesting in https://github.com/puppetlabs/puppet/pull/8672/files#diff-51e52bffe1de00c153bde14dea475f15abdcb6a1fdaf2691633dad8238e654cfR563

Then have the agent send the parameter when calling `Puppet::Resource::Catalog.indirection.find`

Then add the `default_agent_environment` setting you suggested earlier:https://github.com/puppetlabs/puppet/commit/2bf802541d1f2da0cfe505af10e2724d822e8757#diff-bcf3e5c264a570b4d4cfb510d64f91ec1e4fe6b8618c83346781aef92eb767a1R1372-R1377

And update the catalog terminus check to fallback if the requested environment doesn't exist:

```ruby
def find(..)
  env = Puppet.lookup(:environments).get(request.environment)
  if request.options[:check_environment]
    Puppet.debug { ... }
    # if the fallback doesn't exist, then server is misconfigured, so use `get!`
    env = Puppet.lookup(:environments).get!(Puppet[:default_agent_environment])
  else
    return nil
  end
  ...
```